### DISCUSSION TITLE
OTBR - WSL2 Docker Support (With virtualised Ports)

### DISCUSSION THREAD

**Discussion by HamishBanham** [openthread/openthread#8313] [Q&A] [answered]
Hi,

I am looking into creating a OTBR however, due to supply issues I am unable to get hold of a Raspbery Pi that'll meet the minimum requirements laid out by the OTBR documentation. As I don't need the thread network exposed fully, just locally into either the windows 10/11 machine or a WSL2 Ubuntu distro, I am looking into using the OTBR docker container provided within WSL2 Ubuntu Focal. Being able to use a virtualised WSL2 container would simplify the barrier of entry in developing thread/matter products as network interactions could be displayed through a simple web GUI exposed via local port.

The container can be built successfully, but fails to start as the ip6table is missing a lot of the typical rule matches (for example pkttype) and fails to generate a rule chain (Also note, as I am running a Focal distro not Bionic, ip6tables did not work correctly until I modified the container to use Focal). From my understanding this is because windows builds a custom kernal that does not expose a lot of low level networking infrastructure. I am unfamiliar with WSL2 networking and ip6tables in general.

* Can the firewall rules be ignored if the user does not wish to expose the network past the local container?
* Is there a better solution to create a WSL2 based OTBR?
    * Unable to migrate to a full linux distro for various reasons 

Notes: 
* NRF52840 Dongle running the Thread Co-processor (nrf-connect-sdk 2.1.99, USB transport)
* USB exposed using https://learn.microsoft.com/en-us/windows/wsl/connect-usb (Don't know how well this works until I can get OTBR running)

**Comment by jwhui** [comment, marked as answer]
To disable the firewall feature, you can try adding `-DOT_FIREWALL=OFF` to `OTBR_OPTIONS` when building.

Example: https://github.com/openthread/ot-br-posix/blob/b2226038866c6accdf74173bbcbd2b8905a1ce2b/.github/workflows/macOS.yml#L57

**Reply by HamishBanham** [reply]
Hi @jwhui thanks for your reply.

I pulled the docker image from docker hub and modified the `OTBR_OPTIONS` to include `-DOT_FIREWALL=OFF`  (`docker run -p 8080:80 --dns=127.0.0.1 -it --volume /dev/ttyACM0:/dev/ttyACM0 --privileged openthread/otbr --radio-url spinel+hdlc+uart:///dev/ttyACM0 -e OTBR_OPTIONS=-DOT_FIREWALL=OFF`). However. I still run into the issue where the firewall is still being modified:

```
+ firewall_start
+ firewall_stop
+ ip6tables -C FORWARD -o wpan0 -j OTBR_FORWARD_INGRESS
ip6tables v1.6.1: Couldn't load target `OTBR_FORWARD_INGRESS':No such file or directory

Try `ip6tables -h' or 'ip6tables --help' for more information.
+ ip6tables -L OTBR_FORWARD_INGRESS
ip6tables: No chain/target/match by that name.
+ ipset_destroy_if_exist otbr-ingress-deny-src
+ ipset list otbr-ingress-deny-src
ipset v6.34: Kernel support protocol versions 6-7 while userspace supports protocol versions 6-6
The set with the given name does not exist
+ ipset_destroy_if_exist otbr-ingress-deny-src-swap
+ ipset list otbr-ingress-deny-src-swap
ipset v6.34: Kernel support protocol versions 6-7 while userspace supports protocol versions 6-6
The set with the given name does not exist
+ ipset_destroy_if_exist otbr-ingress-allow-dst
+ ipset list otbr-ingress-allow-dst
ipset v6.34: Kernel support protocol versions 6-7 while userspace supports protocol versions 6-6
The set with the given name does not exist
+ ipset_destroy_if_exist otbr-ingress-allow-dst-swap
+ ipset list otbr-ingress-allow-dst-swap
ipset v6.34: Kernel support protocol versions 6-7 while userspace supports protocol versions 6-6
The set with the given name does not exist
+ ipset create -exist otbr-ingress-deny-src hash:net family inet6
+ ipset create -exist otbr-ingress-deny-src-swap hash:net family inet6
+ ipset create -exist otbr-ingress-allow-dst hash:net family inet6
+ ipset create -exist otbr-ingress-allow-dst-swap hash:net family inet6
+ ip6tables -N OTBR_FORWARD_INGRESS
+ ip6tables -I FORWARD 1 -o wpan0 -j OTBR_FORWARD_INGRESS
+ ip6tables -A OTBR_FORWARD_INGRESS -m pkttype --pkt-type unicast -i wpan0 -j DROP
ip6tables: No chain/target/match by that name.
+ die 'Failed to start firewall service'
+ echo ' *** ERROR:  Failed to start firewall service'
 *** ERROR:  Failed to start firewall service
+ exit 1
tail: cannot open '/var/log/syslog' for reading: No such file or directory
tail: no files remaining
```

The issue comes from the `server` shell script calling  `firewall_start` which links to `otbr_firewall->firewall_start`. It does not appear there is a way to disable this from the command line.

**Reply by jwhui** [reply]
`-DOT_FIREWALL=OFF` is a CMake build option and will not work as a run-time option.

**Reply by HamishBanham** [reply]
Noted.

Looks the current container provided won't work for what I'm trying to achieve. For now I'll try and bring up a devcontainer that'll allow the otbr-agent service to run (Right now I am having issues with the transport, this'll need some more investigation on my end). When I have an update, I will update this post.
